clients/da-vinci-client/src/main/java/com/linkedin/davinci/store/rocksdb/RocksDBStorageEngine.java

```diff
@@ -253,4 +254,17 @@ private String getRocksDbEngineConfigPath() {
     return RocksDBUtils.composePartitionDbDir(rocksDbPath, getStoreName(), METADATA_PARTITION_ID) + "/"
         + SERVER_CONFIG_FILE_NAME;
   }
+
+  @Override
+  public boolean hasMemorySpaceLeft() {
+    SstFileManager sstFileManager = factory.getSstFileManager();
+    if (sstFileManager.isMaxAllowedSpaceReached() || sstFileManager.isMaxAllowedSpaceReachedIncludingCompactions()) {
+      return false;
+    }
+    long currentUsage = sstFileManager.getTotalSize();
+    if (factory.getMemoryLimit() - currentUsage >= 2 * factory.getMemtableSize()) {
```

Comment:
Why 2x memtable here?

Reply:
2x just some magic number and we would like to have some headroom before resuming the consumption.
The database stuck can be caused by a flush or a write, and we would like to resume the consumption at least for one flush, and 2 was the number I chose.

Reply:
I think we try to count both current version and future version, so that's the reason?

Reply:
Thanks for the explanation. Could you maybe put a simple comment there to explain the number of 2? Otherwise it could be fairly confusing for other ppl.

Reply:
Sure